### PR TITLE
fix(otel logging): Include ILogger CategoryName in logger output

### LINKED ISSUE #44746: [BUG] OpenTelemetry Exporter for Monitor/App Insights omits LoggerRecord CategoryName
### Library name and version

Azure.Monitor.OpenTelemetry.Exporter 

### Describe the bug

I am in the midst of [switching from App Insights Classic to the OpenTelemetry exporter](https://learn.microsoft.com/en-us/azure/azure-monitor/app/opentelemetry-enable?tabs=aspnetcore), and I found that the Logger exporter omits the `CategoryName` field when sending data to the App Insights endpoint.

The OpenTelemetry model has [LogRecord.CategoryName](https://github.com/open-telemetry/opentelemetry-dotnet/blob/4d254eb2783aa7f1c44053e64dc1d4e261f7eb5b/src/OpenTelemetry/Logs/LogRecord.cs#L172), and the Application Insights classic API faithfully uses the `ILogger.Name` in the `customDimensions/CategoryName` field.

Note that the `CategoryName` field is not listed on this page:
https://opentelemetry.io/docs/specs/otel/logs/data-model/#log-and-event-record-definition
but it IS included in other .NET docs on the opentelemetry site.

This is a pretty severe oversight - I would say the green checkbox here isn't quite true! :)

image

### Expected behavior

When using the OpenTelemetry exporter, calls to `ILogger<T>` pass all the same fields to the App Insights `traces` table as are passed using App Insights classic.

### Actual behavior

The `CategoryName` field is omitted from App Insights by the exporter. I've stepped through the code and verified that it's missing from `LogsHelper.OtelToAzureMonitorLogs()`.

### Reproduction Steps

Run the [OpenTelemetry Exporter demo](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Demo/Program.cs) with an App Insights Connection string, and comment out everything except the Logger part, eg:

```cs
        public static void Main()
        {
            // To use AAD, setup your desired credential and provide to the demo class.
            // var credential = new DefaultAzureCredential();
            // using var traceDemo = new TraceDemo(ConnectionString, credential);

            //using var traceDemo = new TraceDemo(ConnectionString);
            //traceDemo.GenerateTraces();

            //using var metricDemo = new MetricDemo(ConnectionString);
            //metricDemo.GenerateMetrics();

            using var logDemo = new LogDemo(ConnectionString);
            logDemo.GenerateLogs();

            Console.WriteLine("Press any key to exit.");
            Console.ReadLine();
        }
```

then run it, and note the output excludes `CategoryName`:

```jsonc
{
  "name": "Message",
  "time": "2024-06-26T22:22:56.9910123Z",
  "iKey": "02e6ee70-b253-4d75-9785-cf4ba016c9c3",
  "tags": {
    "ai.cloud.role": "[my-namespace]/my-service",
    "ai.cloud.roleInstance": "my-instance",
    "ai.application.ver": "1.0.0-demo",
    "ai.internal.sdkVersion": "dotnet7.0.20:otel1.9.0:ext1.4.0-alpha.20240626"
  },
  "data": {
    "baseType": "MessageData",
    "baseData": {
      "message": "Hello from {name} {price}.",
      "severityLevel": "Information",
      "properties": {
        "name": "tomato",
        "price": "2.99"
      },
      "ver": 2
    }
  }
}
```

### Environment

```
  dotnet --info
.NET SDK:
 Version:           8.0.302
 Commit:            ef14e02af8
 Workload version:  8.0.300-manifests.5273bb1c
 MSBuild version:   17.10.4+10fbfbf2e

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22631
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.302\

Host:
  Version:      8.0.6
  Architecture: x64
  Commit:       3b8b000a0e

.NET SDKs installed:
  6.0.423 [C:\Program Files\dotnet\sdk]
  8.0.302 [C:\Program Files\dotnet\sdk]
```

### PR DESCRIPTION
**Pull request by johncrim**
Previously the `CategoryName` field from the OpenTelemetry `LogRecord` was omitted, which was not consistent with normal ILogger behavior, or with Application Insights classic behavior.

fixes: https://github.com/Azure/azure-sdk-for-net/issues/44746

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs:130]: It does match the Classic ApplicationInsights behavior : https://github.com/microsoft/ApplicationInsights-dotnet/blob/main/LOGGING/src/ILogger/ApplicationInsightsLogger.cs#L159 , but I am not sure if this is the plan going forward! CategoryName is like the instrumentation-scope from OTel, so it should be stored in a more neutral way than "CategoryName", which is very ILogger/.NET specific

**Review by TimothyMothra — CHANGES_REQUESTED** [review]
Hi @johncrim, Thanks for your contribution! I think this is a reasonable approach. 

I'm blocking this PR so we can take some time to discuss this.
We've been mostly focused on mapping OTel concepts to Azure Monitor. "CategoryName" has been missed because it's an ILogger/.NET concept. 
This is on the agenda to discuss in our next group meeting and I hope to have an update next week.

cc: @vishweshbankwar

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs
Comment (johncrim) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs:130]: Well, it seems to me to be a big deal to just drop a key component of the log data just because it's not in the Otel documentation. What would the rationale be for not logging it?

Also `CategoryName` is the property on the `LogRecord` type, so it's at least been implemented in the OpenTelemetry main repo.

I'm not trying to be hostile about this, I just think it's important to include.

Comment (cijothomas) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs:130]: > What would the rationale be for not logging it?

None. It should be logged. But there is no top-level field in TraceTelemetry to put it, so customProperty is the only option. Then it brings the question - what should be the name of the "key".? There is no clear answer as of today. "CategoryName" was used in classic app insights, but that cannot used as-is for OTel. Because "LoggerCategory" is the InstrumentationScope in OTel, so it must be stored consistently across all languages (node/python/java etc.)

> Also CategoryName is the property on the LogRecord type, so it's at least been implemented in the OpenTelemetry main repo.

OTel .NET specially treats `ILogger` and hence it is stored in LogRecord directly. When exporting to OTel's own OTLP Exporter, they are exported as InstrumentationScope only, *not as attributes*.

Comment (johncrim) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs:130]: Thanks for the reply @cijothomas. I looked a little and didn't find a field name in the otel specs/docs for "InstrumentationScope"; and it looks like it's a set of key-value pairs, which if put in a single field would make querying difficult/inefficient. Maybe something like `scope.name` would work, and would align with the conventions.

Comment (cijothomas) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs:130]: > didn't find a field name in the otel specs/docs for "InstrumentationScope";

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-instrumentationscope

https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/common/v1/common.proto#L71

OTel data model has well known way of storing the instrumentationscope. AzureMonitor don't have, and that is why we need to address it (consistently across langauges).

Comment (johncrim) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs:130]: Thanks - I had seen the first doc you referenced (which doesn't provide a schema), but not the protobuf schema (which does). There are 2 parts here that seem relevant:

https://github.com/open-telemetry/opentelemetry-proto/blob/40b3c1b746767cbc13c2e39da3eaf1a23e54ffdd/opentelemetry/proto/logs/v1/logs.proto#L66C1-L81C2

Shows that the formalized InstrumentationScope is within a parent object that also contains all the `LogRecords` - so it's not part of each `LogRecord` in this schema. This schema seems possibly useful for in-proc batching or sending LogRecords over the wire, but not for storing flattened records, so I'll assume it would be flattened into the AI `TraceTelemetry`.

https://github.com/open-telemetry/opentelemetry-proto/blob/40b3c1b746767cbc13c2e39da3eaf1a23e54ffdd/opentelemetry/proto/common/v1/common.proto#L69C1-L81C2

shows the schema for InstrumentationScope .

Since I don't think more radical changes to App Insights storage design are likely, I think using a property `scope.name` is the most practical option. Or another option is to stick with `CategoryName` until a top-level InstrumentationScope map is added (if that seems likely), to avoid multiple schema changes.

Comment (cijothomas) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs:130]: > Since I don't think more radical changes to App Insights storage design are likely, 

I won't rule that out. For OTel Resource, AppInsights has made a totally new design! I don't know if the owners will chose to do similar for InstrumentationScope or not!

> using a property scope.name is the most practical option

Yes, but it is not really my call to make that decision!

**Comment by johncrim** [comment]
I should note that this PR doesn't include any changes to the real-time monitoring client. Once an approach is settled on, it seems likely that the same change should be added there.

**Comment by johncrim** [comment]
Sorry for the hiccup here - I was just trying to rename my branch (the source of this PR) and accidentally closed the PR. Re-opening.

**Review by cijothomas — APPROVED** [review]
This is aligned with the classic application insights sdk behavior : https://github.com/microsoft/ApplicationInsights-dotnet/blob/main/LOGGING/src/ILogger/ApplicationInsightsLogger.cs#L159

I understand there are future plans to better support scope name (across signals), but this is a reasonable balance for now, given how important `ILogger` category is for `ILogger` users.

**Comment by cijothomas** [comment]
@johncrim Could you resolve the merge conflicts? 

I am proposing to merge this and include for the imminent release train, but will let @TimothyMothra make the final decision.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs
Comment (cijothomas) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs:130]: @vishweshbankwar updated that it is decided by the Product team to use the same key as before for now. Resolving this conversation.

**Comment by johncrim** [comment]
@cijothomas  - I haven't looked at the real-time monitoring client much, but I think the same change should be made there. I hadn't done that due to uncertainty around naming.

Taking a look at that work right now...

**Comment by johncrim** [comment]
Hi @cijothomas - I've resolved the merge conflicts in the PR.

I __have not__ made any changes to LiveMetrics. I would be happy to - doesn't look hard - but I would like confirmation that the livemetrics trace document and exception document should have a `CategoryName` property to match the normal telemetry.

**Comment by johncrim** [comment]
I found that `CategoryName` is included in livemetrics using AppInsights classic, so I'm going to add that to this PR.

**Review by vishweshbankwar — APPROVED** [review]
LGTM

**Review by rajkumar-rangaraj — APPROVED** [review]
@johncrim Could you please update the changelog for both exporter and distro?

**Review by TimothyMothra — APPROVED** [review]

**Pull request merged**
